File: python/open_guardrail/guards/content_length_ratio.py
"""Check content vs markup ratio in responses."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult

_TAG_RE = re.compile(r"<[^>]+>")
_WS_RE = re.compile(r"\s+")


class _ContentLengthRatio:
    def __init__(
        self,
        *,
        action: str = "warn",
        min_content_ratio: float = 0.5,
    ) -> None:
        self.name = "content-length-ratio"
        self.action = action
        self._min_ratio = min_content_ratio

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()
        total = len(text)

        if total == 0:
            elapsed = (
                time.perf_counter() - start
            ) * 1000
            return GuardResult(
                guard_name="content-length-ratio",
                passed=False,
                action=self.action,
                score=1.0,
                message="Empty response",
                latency_ms=round(elapsed, 2),
                details={"reason": "empty-response"},
            )

        stripped = _WS_RE.sub(
            " ", _TAG_RE.sub("", text)
        ).strip()
        ratio = len(stripped) / total
        triggered = ratio < self._min_ratio
        score = 1.0 - ratio if triggered else 0.0
        elapsed = (
            time.perf_counter() - start
        ) * 1000

        return GuardResult(
            guard_name="content-length-ratio",
            passed=not triggered,
            action=(
                self.action if triggered else "allow"
            ),
            score=score,
            message=(
                "Low content-to-markup ratio"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details={
                "content_ratio": round(ratio, 2)
            },
        )


def content_length_ratio(
    *,
    action: str = "warn",
    min_content_ratio: float = 0.5,
) -> _ContentLengthRatio:
    return _ContentLengthRatio(
        action=action,
        min_content_ratio=min_content_ratio,
    )

File: packages/guards/src/content-length-ratio.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface ContentLengthRatioOptions {
  action: 'block' | 'warn';
  minContentRatio?: number;
}

const TAG_RE = /<[^>]+>/g;
const WHITESPACE_RE = /\s+/g;

export function contentLengthRatio(options: ContentLengthRatioOptions): Guard {
  const minRatio = options.minContentRatio ?? 0.5;

  return {
    name: 'content-length-ratio',
    version: '0.1.0',
    description: 'Checks content vs markup ratio in responses',
    category: 'format',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();

      const totalLen = text.length;
      if (totalLen === 0) {
        return {
          guardName: 'content-length-ratio',
          passed: false,
          action: options.action,
          score: 1.0,
          latencyMs: Math.round(performance.now() - start),
          details: { reason: 'empty-response' },
        };
      }

      const stripped = text.replace(TAG_RE, '').replace(WHITESPACE_RE, ' ').trim();
      const ratio = stripped.length / totalLen;
      const triggered = ratio < minRatio;
      const score = triggered ? 1.0 - ratio : 0;

      return {
        guardName: 'content-length-ratio',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        latencyMs: Math.round(performance.now() - start),
        details: { contentRatio: Math.round(ratio * 100) / 100 },
      };
    },
  };
}
